compiler/rustc_lint_defs/src/builtin.rs

```diff
@@ -2648,6 +2648,78 @@ declare_lint! {
     };
 }
 
+declare_lint! {
+    /// The `fuzzy_provenance_casts` lint detects an `as` cast between an integer
+    /// and a pointer.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// fn main() {
+    ///     let dangling = 16_usize as *const u8;
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// This lint is part of the strict provenance effort, see [issue #95228].
+    /// Casting an integer to a pointer is considered bad style, as a pointer
+    /// contains, besides the *address* also a *provenance*, indicating what
+    /// memory the pointer is allowed to read/write. Casting an integer, which
+    /// doesn't have provenance, to a pointer requires the compiler to assign
+    /// (guess) provenance. The compiler assigns "all valid" provenance, but
+    /// this penalizes the optimiser and is not well suited for dynamic
+    /// analysis/dynamic program verification (e.g. `miri` or CHERI platforms).
+    ///
+    /// It is much better to use `ptr::with_addr` instead to specify the
+    /// provenance you want.
+    ///
+    /// This might (or might not) become a [future-incompatible] lint or hard
+    /// error in a future rust edition. See [issue #95228] for more details.
+    ///
+    /// [future-incompatible]: ../index.md#future-incompatible-lints
+    /// [issue #95228]: https://github.com/rust-lang/rust/issues/95228
+    pub FUZZY_PROVENANCE_CASTS,
+    Allow,
+    "A fuzzy integer to pointer cast is used",
```

Comment:
Nit: lint descriptions usually start with a lowercase
```suggestion
    "a fuzzy integer to pointer cast is used",
```